### PR TITLE
Add Locations Pre/Post Feature Segment & Segment Admin

### PR DESCRIPTION
**Pull request by madison-boman**
OPS:
- Segments
      -Add
      -Edit
- SegmentTexts
      -Add
      -Edit
      -Delete
- Locations
      -Pre/PostFeatureSegments are settable

PROM:
- Displays on the Location Pages

### CODE REVIEW

File: src/Promenade.Controllers/LocationsController.cs [hunk 1]
```diff
@@ -186,8 +190,49 @@ public async Task<IActionResult> Locations(string locationStub, string featureSt
                     LocationFeatures = new List<LocationsFeaturesViewModel>(),
                     Location = await _locationService.GetLocationByStubAsync(locationStub)
                 };
+                if (viewModel.Location.PreFeatureSegmentId != null)
+                {
+                    var segment = await _segmentService.GetSegmentById(viewModel.Location.PreFeatureSegmentId.Value);
+                    if (segment.IsActive)
```
Comment (haraldnagel) [inline src/Promenade.Controllers/LocationsController.cs]: Business logic (e.g. if the segment is active and is scheduled to be displayed) should happen in the service call and not the controller.

File: src/Promenade.Models/Entities/SegmentText.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+﻿using System.ComponentModel;
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+using Ocuda.Promenade.Models.Abstract;
+
+namespace Ocuda.Promenade.Models.Entities
+{
+    public class SegmentText : BaseEntity
```
Comment (haraldnagel) [inline src/Promenade.Models/Entities/SegmentText.cs]: This should not extend BaseEntity as it should have a composite key of SegmentId and LanguageId (unless this ties in with work @iafb is doing on model refactoring)

File: src/Promenade.Service/Interfaces/Repositories/ISegmentTextRepository.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿
+using Ocuda.Promenade.Models.Entities;
+
+namespace Ocuda.Promenade.Service.Interfaces.Repositories
+{
+    public interface ISegmentTextRepository : IGenericRepository<SegmentText, int>
+    {
+        public SegmentText GetSegmentTextBySgmentId(int segmentId);
```
Comment (haraldnagel) [inline src/Promenade.Service/Interfaces/Repositories/ISegmentTextRepository.cs]: Typo, "SgmentId"

File: src/Promenade.Data/Promenade/SegmentRepository.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+{
+    public class SegmentRepository
+    : GenericRepository<PromenadeContext, Segment, int>, ISegmentRepository
+    {
+        public SegmentRepository(ServiceFacade.Repository<PromenadeContext> repositoryFacade,
+            ILogger<SegmentRepository> logger) : base(repositoryFacade, logger)
+        {
+        }
+
+        public async Task<List<Segment>> GetAllActiveSegments()
```
Comment (haraldnagel) [inline src/Promenade.Data/Promenade/SegmentRepository.cs]: Where is this method called?

File: src/Promenade.Service/Interfaces/Repositories/ISegmentTextRepository.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿
+using Ocuda.Promenade.Models.Entities;
+
+namespace Ocuda.Promenade.Service.Interfaces.Repositories
+{
+    public interface ISegmentTextRepository : IGenericRepository<SegmentText, int>
+    {
+        public SegmentText GetSegmentTextBySegmentId(int segmentId);
```
Comment (haraldnagel) [inline src/Promenade.Service/Interfaces/Repositories/ISegmentTextRepository.cs]: This method is not used. Remove from interface and implementation.

File: src/Promenade.Service/SegmentService.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            ISegmentTextRepository segmentTextRepository)
+            : base(logger, dateTimeProvider)
+        {
+            _segmentRepository = segmentRepository
+                ?? throw new ArgumentNullException(nameof(segmentRepository));
+            _segmentTextRepository = segmentTextRepository
+                ?? throw new ArgumentNullException(nameof(segmentTextRepository));
+        }
+
+        public async Task<Segment> GetSegmentById(int segmentId)
```
Comment (haraldnagel) [inline src/Promenade.Service/SegmentService.cs]: This method is not used. Remove.

File: src/Promenade.Service/SegmentService.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            var segment = await _segmentRepository.FindAsync(segmentId);
+            segment.SegmentText = _segmentTextRepository.GetSegmentTextBySegmentId(segment.Id);
+            return segment;
+        }
+
+        public async Task<SegmentText> GetSegmentTextBySegmentAndLanguageId(int segmentId, int languageId)
+        {
+            var segment = await _segmentRepository.FindAsync(segmentId);
+            if (segment != null && segment.IsActive)
```
Comment (haraldnagel) [inline src/Promenade.Service/SegmentService.cs]: `if (segment?.IsActive == true)`

File: src/Promenade.Service/SegmentService.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            {
+                if (segment.StartDate != null && segment.EndDate != null)
+                {
+                    var today = DateTime.Now;
```
Comment (haraldnagel) [inline src/Promenade.Service/SegmentService.cs]: Never use `DateTime.Now`, always `_dateTimeProvider.Now`; also don't use a today variable for this, just put `_dateTimeProvider.Now` directly in the if twice.

File: src/Promenade.Service/SegmentService.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+                    if (today > segment.StartDate && today < segment.EndDate)
```
Comment (haraldnagel) [inline src/Promenade.Service/SegmentService.cs]: I'm confused about this logic. What if a StartDate is set but EndDate is null? Then this logic will not work.

File: src/Ops.Controllers/Areas/SiteManagement/LocationsController.cs [hunk 10]
```diff
@@ -242,8 +242,7 @@ public async Task<IActionResult> EditLocation(Location location)
                 {
                     var oldLocation = await _locationService.GetLocationByIdAsync(location.Id);
                     if (!(oldLocation.Address.Equals(location.Address) && oldLocation.City.Equals(location.City)
-                        && oldLocation.State.Equals(location.State) && oldLocation.Zip.Equals(location.Zip)
-                        && oldLocation.Country.Equals(location.Country)))
+                        && oldLocation.State.Equals(location.State) && oldLocation.Zip.Equals(location.Zip)))
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/Ops.Controllers/Areas/SiteManagement/LocationsController.cs]: My fault. I removed this just when I was testing and kept reinserting the SQL script data. Just forgot to put it back in.

File: src/Ops.Data/Ops/SegmentTextRepository.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+        public SegmentText GetSegmentTextBySegmentId(int segmentId, int languageId)
+        {
+            return DbSet
+                .AsNoTracking()
+                .Where(_ => _.SegmentId == segmentId && _.LanguageId == languageId)
+                .FirstOrDefault();
+        }
+
+        public async Task<List<string>> GetLanguageById(int segmentId)
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/Ops.Data/Ops/SegmentTextRepository.cs]: Does 'GetUsedLanguageNamesBySegmentId' make more sense?

File: src/Ops.Service/SegmentService.cs [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            return await _segmentRepository.GetPaginatedListAsync(filter);
+        }
+
+        public async Task<Segment> GetSegmentById(int segmentId)
+        {
+            return await _segmentRepository.FindAsync(segmentId);
+        }
+
+        public SegmentText GetSegmentTextBySegmentLanguageId(SegmentText segmentText)
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/Ops.Service/SegmentService.cs]: There isn't a reason to have both when they do the same thing. Removing the line 43 function

File: src/Promenade.Controllers/LocationsController.cs [hunk 13]
```diff
@@ -18,14 +18,20 @@ namespace Ocuda.Promenade.Controllers
...
 
         public static string Name { get { return "Locations"; } }
 
         public LocationsController(ServiceFacades.Controller<LocationsController> context,
-            LocationService locationService) : base(context)
+            LocationService locationService, SegmentService segmentService, LanguageService languageService ) : base(context)
         {
             _locationService = locationService
                 ?? throw new ArgumentNullException(nameof(locationService));
+            _languageService = languageService
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/Promenade.Controllers/LocationsController.cs]: I have to get the default Language Id from the service to get the correct SegmentText. Is there a function that I could call to do so? Similar to the GetCurrentActiveUserId?

File: src/Promenade.Service/Interfaces/Repositories/ISegmentTextRepository.cs (hunk 5 above)
Comment (madison-boman) [reply src/Promenade.Service/Interfaces/Repositories/ISegmentTextRepository.cs]: Removed

File: src/Promenade.Service/SegmentService.cs (hunk 6 above)
Comment (madison-boman) [reply src/Promenade.Service/SegmentService.cs]: Removed

File: src/Promenade.Service/SegmentService.cs (hunk 7 above)
Comment (madison-boman) [reply src/Promenade.Service/SegmentService.cs]: Updated 'if' statement

File: src/Promenade.Service/SegmentService.cs (hunk 8 above)
Comment (madison-boman) [reply src/Promenade.Service/SegmentService.cs]: Noted for future use! Updating

File: src/Promenade.Service/SegmentService.cs (hunk 9 above)
Comment (madison-boman) [reply src/Promenade.Service/SegmentService.cs]: Revised Logic to handle this case

File: src/Promenade.Models/Entities/SegmentText.cs [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+        [Key]
+        [Required]
+        [DisplayName("Language")]
+        public int LanguageId { get; set; }
+
+        public Language Language { get; set; }
+
+        public string Header { get; set; }
+
+        public string Text { get; set; }
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/Promenade.Models/Entities/SegmentText.cs]: Updated

File: src/Promenade.Models/Entities/SegmentText.cs [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+        public Segment Segment { get; set; }
+
+        [Key]
+        [Required]
+        [DisplayName("Language")]
+        public int LanguageId { get; set; }
+
+        public Language Language { get; set; }
+
+        public string Header { get; set; }
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/Promenade.Models/Entities/SegmentText.cs]: Updated

File: src/Promenade.Models/Entities/Segment.cs [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+{
+    public class Segment : BaseEntity
+    {
+        public DateTime? StartDate { get; set; }
+
+        public DateTime? EndDate { get; set; }
+
+        public bool IsActive { get; set; }
+
+        public string Name { get; set; }
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/Promenade.Models/Entities/Segment.cs]: Updated

File: src/Promenade.Data/Promenade/SegmentTextRepository.cs [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+        public SegmentText GetSegmentTextBySegmentId(int segmentId)
+        {
+            return DbSet
+                .AsNoTracking()
+                .Where(_ => _.SegmentId == segmentId)
+                .FirstOrDefault();
+        }
+
+        public SegmentText GetSegmentTextBySgmentAndLanguageId(int segmentId, int languageId)
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/Promenade.Data/Promenade/SegmentTextRepository.cs]: This was a copy paste function. I corrected the one from Ops but, did realize it was in Prom.

File: src/Promenade.Controllers/LocationsController.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/Promenade.Controllers/LocationsController.cs]: Resolved

File: src/Promenade.Controllers/LocationsController.cs [hunk 18]
```diff
@@ -186,8 +192,26 @@ public async Task<IActionResult> Locations(string locationStub, string featureSt
                     LocationFeatures = new List<LocationsFeaturesViewModel>(),
                     Location = await _locationService.GetLocationByStubAsync(locationStub)
                 };
+                if (viewModel.Location.PreFeatureSegmentId.HasValue)
+                {
+                    viewModel.PreFeatureSegment = await _segmentService.GetSegmentTextBySegmentAndLanguageId(
+                        viewModel.Location.PreFeatureSegmentId.Value, await _languageService.GetDefaultLanguageIdAsync());
```
[the remark this answers is not in the document]
Comment (madison-boman) [reply src/Promenade.Controllers/LocationsController.cs]: Resolved

**Pull request merged**